torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -200,6 +200,15 @@ FullOp::FullOp(const FullOp* src, IrCloner* ir_cloner)
       dtype_(src->dtype()),
       fill_value_(ir_cloner->clone(src->fill_value_)) {}
 
+Expr* FullOp::shallowCopy() const {
+  auto result = IrBuilder::create<FullOp>(output(0), fill_value_, dtype_);
+  if (container()->isA<kir::Kernel>()) {
+    result->setPredicate(predicate());
+    result->setWritePredicate(writePredicate());
```

Comment:
These two lines are repeated many times. Is there any way to avoid the repetition? Looks like this part could be done at the `Expr` class, so some inheritance may help? Why not a copy constructor?

Reply:
We don't have copy constructor because we need an`IrBuilderPasskey` to create objects...

Reply:
It doesn't need to be a copy constructor. I'm just concerned as they are repeated many times. Maybe adding an implementation of `Expr::shallowCopy` as a protected function and move `setPredicate` and `setWritePredicate` calls to it? This way, all the subclasses would just need to call it. They still need to explicitly call it, but still it'd be much easier when we would need to do any change with `setPredicate` and `writePredicate`.

Reply:
I added `Expr::copyPredicatesFrom` for this purpose. I don't want to add `Expr::shallowCopy`, because I want the compiler to be able to tell me that I forget to override `shallowCopy` when adding a new expression type (this error do happens a few times when working on this PR).

Reply:
That sounds even better. Thanks!